src/ethereum_spec_tools/forks.py

```diff
@@ -5,23 +5,115 @@
 Detects Python packages that specify Ethereum hardforks.
 """
 
+import functools
 import importlib
 import pkgutil
 from pkgutil import ModuleInfo
 from types import ModuleType
-from typing import Any, Iterator, List, Optional, Type, TypeVar
+from typing import Any, Dict, Iterator, List, Optional, Type, TypeVar
 
 import ethereum
 
 H = TypeVar("H", bound="Hardfork")
 
 
+@functools.total_ordering
+class ForkCriteria:
```

Comment:
Ah. I missed the mark on explaining this I think. Or maybe my idea was terrible :rofl:

I was thinking we'd have:

```python
# src/ethereum/base_types.py (maybe)

class HeaderThing(Protocol):
    number: Uint
    timestamp: U256

class ForkActivation(ABC):
    @abstractmethod
    def activate_for(self, header: HeaderThing) -> bool:
        pass

class ByNumber(ForkActivation):
    def __init__(self, number: Uint):
        self.number = number

    def activate_for(self, header: HeaderThing) -> bool:
        return header.number >= self.number

class ByTimestamp(ForkActivation):
    def __init__(self, timestamp: U256):
        self.timestamp = timestamp

    def activate_for(self, header: HeaderThing) -> bool:
        return header.timestamp >= self.timestamp
```

```python
# src/ethereum/paris/__init__.py

MAINNET_ACTIVATION = ByTimestamp(1681338455)
```

Reply:
Now that I've written it up, I am not so certain it's a good fit. It's very object oriented. What are your thoughts?

Reply:
I disagree with changing the `MAINNET_FORK_BLOCK` and `MAINNET_FORK_TIMESTAMP` constants. Forks exist in almost total ignorance of anything that happens outside of them. We consistently adopt hacky behaviour in `ethereum_spec_tools` to work around this and I don't see why we should change that policy here.

I think we need a `ForkCriteria` class with the same semantics as the following Rust type:
```rust
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
enum ForkCriteria {
    BlockNumber(u64),
    Timestamp(u64),
    Unscheduled,
}
```

`Ord` is required so that we can easily sort forks by activation order. Implementing that type clearly and succinctly in Python is surprisingly hard. I think your approach would work if we didn't need the ordering.

Reply:
> I disagree with changing the `MAINNET_FORK_BLOCK` and `MAINNET_FORK_TIMESTAMP` constants. [...] We consistently adopt hacky behaviour in `ethereum_spec_tools` to work around this and I don't see why we should change that policy here.

Hm, so you find `MAINNET_ACTIVATION = ByTimestamp(1681338455)` less readable than `MAINNET_FORK_TIMESTAMP = 1681338455`? Or maybe that it's adding too much complexity to the spec?

Personally I think we can change the spec to make it more compatible with the tooling, but _only as long as it doesn't make understanding the spec more difficult_. If we can avoid a hacky solution in `ethereum_spec_tools` **and** keep the spec understandable, I'd be the happiest.

> I think we need a `ForkCriteria` class with the same semantics as the following Rust type [...]

That's exactly what I was trying to replicate in Python, but with classes instead of an enum.

> I think your approach would work if we didn't need the ordering.

Could we add `@abstractmethod` magic methods (eg. `__lt__`) to `ForkActivation` to implement ordering?